config/pkcs11/config.go

```diff
@@ -0,0 +1,116 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package pkcs11config
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/containers/ocicrypt/crypto/pkcs11"
+	enchelpers "github.com/containers/ocicrypt/helpers"
+	"gopkg.in/yaml.v2"
+)
+
+// OcicryptoConfig represents the format of an imgcrypt.conf config file
+type OcicryptConfig struct {
+	Pkcs11Config pkcs11.Pkcs11Config `yaml:"pkcs11"`
+}
+
+const CONFIGFILE = "ocicrypt-pkcs11.conf"
+
+// parseConfigFile parses a configuration file; it is not an error if the configuration file does
+// not exist, so no error is returned.
+func parseConfigFile(filename string) (*OcicryptConfig, error) {
+	// a non-existent config file is not an error
+	_, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return nil, nil
+	}
+
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+
+	ic := &OcicryptConfig{}
+	err = yaml.Unmarshal(data, ic)
+	return ic, err
+}
+
+// getConfiguration tries to read the configuration file at the following locations
+// 1) ${OCICRYPT_PKCS11_CONFIG}
+// 2) ${XDG_CONFIG_HOME}/ocicrypt-pkcs11.conf
+// 3) ${HOME}/.config/ocicrypt-pkcs11.conf
+// 4) /etc/ocicrypt-pkcs11.conf
+// If no configuration file could be found or read a null pointer is returned
+func getConfiguration() (*OcicryptConfig, error) {
+	filename := os.Getenv("OCICRYPT_PKCS11_CONFIG")
+	if len(filename) > 0 {
+		ic, err := parseConfigFile(filename)
+		if err != nil || ic != nil {
+			return ic, nil
+		}
+	}
+	envvar := os.Getenv("XDG_CONFIG_HOME")
+	if len(envvar) > 0 {
+		ic, err := parseConfigFile(path.Join(envvar, CONFIGFILE))
+		if err != nil || ic != nil {
+			return ic, nil
+		}
+	}
+	envvar = os.Getenv("HOME")
+	if len(envvar) > 0 {
+		ic, err := parseConfigFile(path.Join(envvar, ".config", CONFIGFILE))
+		if err != nil || ic != nil {
+			return ic, nil
+		}
+	}
+	return parseConfigFile(path.Join("etc", CONFIGFILE))
+}
+
+// getDefaultCryptoConfigOpts returns default crypto config opts needed for pkcs11 module access
+func getDefaultCryptoConfigOpts() (enchelpers.CryptoConfigOpts, error) {
+	mdyaml := pkcs11.GetDefaultModuleDirectoriesYaml()
+	config := fmt.Sprintf("module-directories:\n"+
+		" %s"+
+		"allowed-module-paths:\n"+
+		" %s", mdyaml, mdyaml)
+	p11conf, err := pkcs11.ParsePkcs11ConfigFile([]byte(config))
+	if err != nil {
+		return enchelpers.CryptoConfigOpts{}, err
+	}
+	return enchelpers.CryptoConfigOpts{
+		Pkcs11Config: p11conf,
+	}, nil
+}
+
+// GetCryptoConfigOpts gets the CryptoConfigOpts either from a configuration file or if none is
+// found the default ones are returned
+func GetCryptoConfigOpts() (enchelpers.CryptoConfigOpts, error) {
+	ic, err := getConfiguration()
+	if err != nil {
+		return enchelpers.CryptoConfigOpts{}, err
+	}
+	if ic == nil {
+		return getDefaultCryptoConfigOpts()
```

Comment:
suggestion: get pkcs11Config here and then return all at once at the bottom

Reply:
The question is now whether to use an internal policy at all.